packages/nanocurrency/src/assembly/blake2b.ts

```diff
@@ -0,0 +1,167 @@
+import { HEADER_SIZE } from 'internal/arraybuffer'
+
+export class Context {
+  b: Uint8Array;
+  h: Uint64Array;
+  t: Uint64Array;
+  c: usize;
+  outlen: usize;
+}
+
+const HASH_LENGTH: u32 = 8;
+
+function B2B_GET64(arr: Uint8Array, i: u32): u64 {
+  return load<u64>(changetype<usize>(arr.buffer) + arr.byteOffset + i, HEADER_SIZE);
+}
+
+function B2B_G(a: i32, b: i32, c: i32, d: i32, x: u64, y: u64): void {
+  v[a] = v[a] + v[b] + x;
+  v[d] = rotr<u64>(v[d] ^ v[a], 32);
+  v[c] = v[c] + v[d];
+  v[b] = rotr<u64>(v[b] ^ v[c], 24);
+  v[a] = v[a] + v[b] + y;
+  v[d] = rotr<u64>(v[d] ^ v[a], 16);
+  v[c] = v[c] + v[d];
+  v[b] = rotr<u64>(v[b] ^ v[c], 63);
+}
+
+let BLAKE2B_IV_DATA: u64[] = [
+  0x6a09e667f3bcc908,
+  0xbb67ae8584caa73b,
+  0x3c6ef372fe94f82b,
+  0xa54ff53a5f1d36f1,
+  0x510e527fade682d1,
+  0x9b05688c2b3e6c1f,
+  0x1f83d9abfb41bd6b,
+  0x5be0cd19137e2179,
+];
+
+let BLAKE2B_IV = new Uint64Array(BLAKE2B_IV_DATA.length);
+for (let i = 0; i < BLAKE2B_IV.length; i++) {
+  BLAKE2B_IV[i] = BLAKE2B_IV_DATA[i];
+}
+
+const SIGMA: u8[][] = [
+  [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15],
+  [14, 10, 4, 8, 9, 15, 13, 6, 1, 12, 0, 2, 11, 7, 5, 3],
+  [11, 8, 12, 0, 5, 2, 15, 13, 10, 14, 3, 6, 7, 1, 9, 4],
+  [7, 9, 3, 1, 13, 12, 11, 14, 2, 6, 5, 10, 4, 0, 15, 8],
+  [9, 0, 5, 7, 2, 4, 10, 15, 14, 1, 11, 12, 6, 8, 3, 13],
+  [2, 12, 6, 10, 0, 11, 8, 3, 4, 13, 7, 5, 15, 14, 1, 9],
+  [12, 5, 1, 15, 14, 13, 4, 10, 0, 7, 6, 3, 9, 2, 8, 11],
+  [13, 11, 7, 14, 12, 1, 3, 9, 5, 0, 15, 4, 8, 6, 2, 10],
+  [6, 15, 14, 9, 11, 3, 0, 8, 12, 2, 13, 7, 1, 4, 10, 5],
+  [10, 2, 8, 4, 7, 6, 1, 5, 15, 11, 9, 14, 3, 12, 13, 0],
+  [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15],
+  [14, 10, 4, 8, 9, 15, 13, 6, 1, 12, 0, 2, 11, 7, 5, 3],
+];
```

Comment:
Also flattening `SIGMA` could be potentially better that two dimensional array, but this need speed measurement